typegate/src/typegate/artifacts/shared.ts

```diff
@@ -0,0 +1,255 @@
+      dir: STORE_TEMP_DIR,
+    });
+    const file = await Deno.open(tmpFile, {
+      write: true,
+      truncate: true,
+    });
+
+    const hasher = createHash("sha256");
+    await stream
+      .pipeThrough(new HashTransformStream(hasher))
+      .pipeTo(file.writable);
+
+    const hash = hasher.digest("hex");
+
+    const readFile = await Deno.open(tmpFile, { read: true });
+    // Read file content into a Uint8Array
+    const fileContent = await readAll(readFile);
+    console.log(`Persisting artifact to S3`);
+    const _ = await this.#s3.putObject({
+      Bucket: this.#syncConfig.s3Bucket,
+      Body: fileContent,
+      Key: resolveS3Key(hash),
+    });
+    readFile.close();
+
+    return hash;
+  }
+
+  override async delete(hash: string): Promise<void> {
+    const _ = await this.#s3.deleteObject({
+      Bucket: this.#syncConfig.s3Bucket,
+      Key: resolveS3Key(hash),
+    });
+  }
+
+  override async has(hash: string): Promise<boolean> {
+    try {
+      const _ = await this.#s3.headObject({
+        Bucket: this.#syncConfig.s3Bucket,
+        Key: resolveS3Key(hash),
+      });
+      return true;
+    } catch {
+      return false;
+    }
+  }
+
+  override async getLocalPath(
+    meta: ArtifactMeta,
+    deps?: ArtifactMeta[] | undefined,
+  ): Promise<string> {
+    for (const dep of deps ?? []) {
+      await this.#downloadFromRemote(dep.hash, dep.relativePath);
+      await getLocalPath(dep);
+    }
+
+    await this.#downloadFromRemote(meta.hash, meta.relativePath);
+    return await getLocalPath(meta);
+  }
+
+  async #downloadFromRemote(hash: string, relativePath: string) {
+    if (await this.#existsLocally(hash)) {
+      return;
+    }
+    const targetFile = resolve(STORE_DIR, hash);
+    const response = await this.#s3.getObject({
+      Bucket: this.#syncConfig.s3Bucket,
+      Key: resolveS3Key(hash),
+    });
+
+    if (response.$metadata.httpStatusCode === 404) {
+      throw new Error(
+        `Artifact ${relativePath} with hash ${relativePath} not found`,
+      );
+    }
+
+    if (response.Body) {
+      const file =
+        (await Deno.open(targetFile, { write: true, create: true })).writable;
+      await response.Body.transformToWebStream().pipeTo(file);
+    } else {
+      throw new Error(`Failed to download artifact ${relativePath} from s3`);
+    }
+  }
+
+  async #existsLocally(hash: string) {
+    try {
+      await Deno.stat(resolve(STORE_DIR, hash));
+      return true;
+    } catch {
+      return false;
+    }
+  }
+
+  override async prepareUpload(
+    meta: ArtifactMeta,
+    origin: URL,
+  ): Promise<string | null> {
+    // should not be uploaded again
+    if (await this.has(ArtifactStore.getArtifactKey(meta))) {
+      return null;
+    }
+    const [url, expirationTime] = await ArtifactStore.createUploadUrl(
+      origin,
+      meta.typegraphName,
+    );
+    await this.#addUrlToRedis(url, serializeToRedisValue(meta), expirationTime);
```

Comment:
Not the right expiration time here, EXPIRE/PEXPIRE expects a duration, not an "instant"... You can directly pass the 5 min value... It could be configurable though...
Redis also has EXPIREAT/PEXPIREAT...